### PR TITLE
Add vmap rules for multi_margin_loss

### LINKED ISSUE #139241: Please implement batching rule for torch.nn.functional.multi_margin_loss
### 🚀 The feature, motivation and pitch

I'm working on a metric to measure the efficiency of a neural network. My approach requires computing per-example gradient using multi_margin_loss.
Like evaluating the test accuracy of a model, the more examples used, the better the estimation. I would like to measure the efficiency of a lot of models using their entire validation set, so a speed boost would save so much time.

### Alternatives

In fact, my original idea is really simple, the loss is simply the output with the additive inverse of argmax value:
```
  m = torch.ones_like(output)
  m[label] = -1
  loss = output * m + 2
```
I've tried every way I know (indexing, scatter, one_hot, arange) to do this: `m[label] = -1`. However, I always run into problems with indexing BatchedTensor.

If vmap apply a function over a batch of data, I think a per-example function shouldn't have to deal with BatchedTensor in the first place. I'm new to all of this so I don't know if I made a mistake, or what is needed in this case, maybe the ability to index, or batching rule for scatter? 

I don't know which is the solution, but I feel like solving this gives more flexibility than implementing batching rule for any other function.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by vineeshah**
Fixes #139241

Adds batching rules for `aten::multi_margin_loss` and
`aten::multi_margin_loss_backward`.

Previously, using `vmap` with `torch.nn.functional.multi_margin_loss`
or per-example gradients through it could hit missing batching rule warnings.
This patch handles batched `self`/`target` inputs by moving the vmap batch
dimension to the front, flattening the logical input into the native loss
shape, calling the existing `multi_margin_loss` implementation with
`Reduction::None`, then restoring the expected logical output shape and
applying reduction without reducing over the vmap dimension.

The backward rule mirrors the same logical flattening and calls the native
`multi_margin_loss_backward` implementation.

This intentionally does not add support for batching over the `weight`
argument. `weight=None` and shared/unbatched `weight` are supported.

Test plan:
- `python3 test/functorch/test_ops.py -k multi_margin_loss`
- `python3 test/functorch/test_vmap.py -k multi_margin_loss`
- `python3 test/functorch/test_vmap_registrations.py`
- `python3 /Users/vineet/projects/issues/repro_pytorch_139241_vmap_multi_margin_loss.py`
- `git diff --check`

cc @Chillee @samdow @kshitij12345

### CODE REVIEW

File: aten/src/ATen/functorch/BatchRulesLoss.cpp [hunk 1]
```diff
@@ -74,6 +74,155 @@ smooth_l1_loss_batch_rule(const at::Tensor& self, std::optional<int64_t> self_bd
...
+    std::optional<int64_t> weight_bdim,
+    int64_t reduction) {
+  TORCH_CHECK(
+      !weight_bdim.has_value(),
+      "vmap: batching over the weight argument of multi_margin_loss is not supported");
+
+  const auto bdim_size = get_bdim_size2(self, self_bdim, target, target_bdim);
+  auto [self_flat, target_flat, nframe, target_has_logical_dim, self_logical_sizes] =
+      multi_margin_loss_prepare_inputs(self, self_bdim, target, target_bdim, bdim_size);
+  (void)self_logical_sizes;
```
Comment (Skylion007) [inline aten/src/ATen/functorch/BatchRulesLoss.cpp]: Why? Just use _ or std::ignore in the structure bindings above

File: aten/src/ATen/functorch/BatchRulesLoss.cpp [hunk 2]
```diff
@@ -74,6 +74,155 @@ smooth_l1_loss_batch_rule(const at::Tensor& self, std::optional<int64_t> self_bd
...
+    int64_t reduction) {
+  TORCH_CHECK(
+      !weight_bdim.has_value(),
+      "vmap: batching over the weight argument of multi_margin_loss_backward is not supported");
+
+  const auto bdim_size = get_bdim_size3(
+      grad_output, grad_output_bdim, self, self_bdim, target, target_bdim);
+  auto [self_flat, target_flat, nframe, target_has_logical_dim, self_logical_sizes] =
+      multi_margin_loss_prepare_inputs(self, self_bdim, target, target_bdim, bdim_size);
+  (void)target_has_logical_dim;
```
Comment (Skylion007) [inline aten/src/ATen/functorch/BatchRulesLoss.cpp]: Same here

**Review by Skylion007 — APPROVED** [review]

**Comment by Skylion007** [comment]
@claude review this PR

File: aten/src/ATen/functorch/BatchRulesLoss.cpp [hunk 3]
```diff
@@ -74,6 +74,155 @@ smooth_l1_loss_batch_rule(const at::Tensor& self, std::optional<int64_t> self_bd
...
+    const Tensor& target,
+    std::optional<int64_t> target_bdim,
+    int64_t bdim_size) {
+  auto self_ = moveBatchDimToFront(self, self_bdim);
+  auto target_ = moveBatchDimToFront(target, target_bdim);
+  self_ = ensure_has_bdim(self_, self_bdim.has_value(), bdim_size);
+  target_ = ensure_has_bdim(target_, target_bdim.has_value(), bdim_size);
+
+  const auto self_logical_rank = rankWithoutBatchDim(self, self_bdim);
+  TORCH_CHECK(
```
Comment (Skylion007) [inline aten/src/ATen/functorch/BatchRulesLoss.cpp]: I guess technically should be TORCH_CHECK_VALUE

**Review by Skylion007 — CHANGES_REQUESTED** [review]
Nits on the void args

File: aten/src/ATen/functorch/BatchRulesLoss.cpp [hunk 4]
```diff
@@ -74,6 +74,155 @@ smooth_l1_loss_batch_rule(const at::Tensor& self, std::optional<int64_t> self_bd
...
+    bool target_has_logical_dim,
+    int64_t reduction) {
+  auto result_ = result.reshape({bdim_size, nframe});
+  if (reduction == Reduction::None) {
+    return target_has_logical_dim ? result_ : result_.squeeze(1);
+  }
+  if (nframe == 1) {
+    return result_.squeeze(1);
+  }
+  if (reduction == Reduction::Sum) {
```
Comment (Skylion007) [inline aten/src/ATen/functorch/BatchRulesLoss.cpp]: Code duplication, you can add edit apply_loss_reduction right below?

**Comment by Skylion007** [comment]
Could you deal with the comment to see if you can support a batched weight too? Similar losses support it

**Comment by vineeshah** [comment]
sounds good, let me make the necessary changes.

**Comment by vineeshah** [comment]
sounds good, let me make the necessary changes.

File: aten/src/ATen/functorch/BatchRulesLoss.cpp (hunk 3 above)
Comment (vineeshah) [reply aten/src/ATen/functorch/BatchRulesLoss.cpp]: yeah correct

File: aten/src/ATen/functorch/BatchRulesLoss.cpp (hunk 4 above)
Comment (vineeshah) [reply aten/src/ATen/functorch/BatchRulesLoss.cpp]: yeah i could but needed another arg for dim

File: aten/src/ATen/functorch/BatchRulesLoss.cpp (hunk 1 above)
Comment (vineeshah) [reply aten/src/ATen/functorch/BatchRulesLoss.cpp]: using [[maybe unused]] now to keep it in one line

**Comment by vineeshah** [comment]
> Could you deal with the comment to see if you can support a batched weight too? Similar losses support it

Batched weight isn't covered by the OpInfo vmap tests (kwargs aren't batched), so this would be a manually-tested-only path. Happy to add it if you want parity with BCE - should I, or is a clearer error message acceptable for now?

**Comment by Skylion007** [comment]
> > Could you deal with the comment to see if you can support a batched weight too? Similar losses support it
> 
> 
> 
> Batched weight isn't covered by the OpInfo vmap tests (kwargs aren't batched), so this would be a manually-tested-only path. Happy to add it if you want parity with BCE - should I, or is a clearer error message acceptable for now?

Parity would be good. @claude how would you enable this test case in opinfo?

**Comment by vineeshah** [comment]
I should've mentioned it but the latest commit already handles it. Please check it out, I wrote a test too.

**Comment by zou3519** [comment]
@claude review this please and put the review inside a GitHub collapsible section
